From 8ea58cad1dd5455feb36ea7406017bf19679825f Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:52:41 -0400
Subject: [tools] lib/traceevent: Unregister handler when sched_switch plugin is unloaded

Message-id: <1408953210-26343-274-git-send-email-jolsa@redhat.com>
Patchwork-id: 88889
O-Subject: [PATCH RHEL7.1 BZ1133083 273/322] tools lib traceevent: Unregister handler when sched_switch plugin is unloaded
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 02bafd377c1137d0705f224881cd21de123204f0
Author: Namhyung Kim <namhyung@kernel.org>
Date: Thu Jan 16 11:31:13 2014 +0900

description
===========
The event handlers should be unregistered when the plugin is unloaded
otherwise they'll try to access invalid memory.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/lib/traceevent/plugin_sched_switch.c b/tools/lib/traceevent/plugin_sched_switch.c
index fea3724..f1ce600 100644
--- a/tools/lib/traceevent/plugin_sched_switch.c
+++ b/tools/lib/traceevent/plugin_sched_switch.c
@@ -146,3 +146,15 @@ int PEVENT_PLUGIN_LOADER(struct pevent *pevent)
           sched_wakeup_handler, NULL);
  return 0;
 }
+
+void PEVENT_PLUGIN_UNLOADER(struct pevent *pevent)
+{
+ pevent_unregister_event_handler(pevent, -1, "sched", "sched_switch",
+     sched_switch_handler, NULL);
+
+ pevent_unregister_event_handler(pevent, -1, "sched", "sched_wakeup",
+     sched_wakeup_handler, NULL);
+
+ pevent_unregister_event_handler(pevent, -1, "sched", "sched_wakeup_new",
+     sched_wakeup_handler, NULL);
+}
-- 
1.7.1